File: clients/OpenClawClient.cs
using System;
using System.Collections.Concurrent;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using System.Net.WebSockets;
using System.Collections.Generic;

namespace OpenClawClient
{
    public class OpenClawClient : IDisposable
    {
        private readonly HttpClient _http;
        private readonly string _baseUrl;
        private readonly string _token;
        
        private ClientWebSocket _ws;
        private CancellationTokenSource _wsCts;
        private readonly ConcurrentDictionary<string, List<Action<object>>> _wsCallbacks = new();
        private readonly BlockingCollection<object> _wsMessageQueue = new();
        private bool _wsConnected;

        public bool IsWsConnected => _wsConnected;

        public OpenClawClient(string baseUrl = "http://localhost:8787", string token = null)
        {
            _baseUrl = baseUrl.TrimEnd('/');
            _token = token;
            _http = new HttpClient();
            
            if (!string.IsNullOrEmpty(_token))
            {
                _http.DefaultRequestHeaders.Add("Authorization", $"Bearer {_token}");
            }
            
            InitCallbacks();
        }

        private void InitCallbacks()
        {
            foreach (var type in new[] { "hello", "init", "progress", "done", "error", "pong" })
            {
                _wsCallbacks[type] = new List<Action<object>>();
            }
        }

        // ============ REST API ============

        public async Task<HealthResponse> GetHealthAsync()
        {
            return await GetAsync<HealthResponse>("/api/health");
        }

        public async Task<ModelsResponse> GetModelsAsync()
        {
            return await GetAsync<ModelsResponse>("/api/models");
        }

        public async Task<SkillsResponse> GetSkillsAsync()
        {
            return await GetAsync<SkillsResponse>("/api/skills");
        }

        public async Task<HooksResponse> GetHooksAsync()
        {
            return await GetAsync<HooksResponse>("/api/hooks");
        }

        public async Task<SessionsResponse> GetSessionsAsync()
        {
            return await GetAsync<SessionsResponse>("/api/sessions");
        }

        public async Task<SessionResponse> GetSessionAsync(string sessionId)
        {
            return await GetAsync<SessionResponse>($"/api/sessions/{sessionId}");
        }

        public async Task<SessionResponse> CreateSessionAsync(string title, object settings = null)
        {
            var body = new { title, settings };
            return await PostAsync<SessionResponse>("/api/sessions", body);
        }

        public async Task<RunResponse> RunAsync(string modelId, string objective, 
            string sessionId = null, string[] enabledSkillIds = null, object settings = null)
        {
            var body = new { modelId, objective, sessionId, enabledSkillIds, settings };
            return await PostAsync<RunResponse>("/api/run", body);
        }

        public async IAsyncEnumerable<StreamEvent> RunStreamAsync(string modelId, string objective,
            string sessionId = null, string[] enabledSkillIds = null, object settings = null)
        {
            var body = new { modelId, objective, sessionId, enabledSkillIds, settings };
            var json = JsonSerializer.Serialize(body);
            var content = new StringContent(json, Encoding.UTF8, "application/json");

            var request = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/api/run/stream") { Content = content };
            if (!string.IsNullOrEmpty(_token))
                request.Headers.Add("Authorization", $"Bearer {_token}");

            var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
            response.EnsureSuccessStatusCode();

            using var stream = await response.Content.ReadAsStreamAsync();
            using var reader = new System.IO.StreamReader(stream);
            
            var buffer = new StringBuilder();
            
            while (!reader.EndOfStream)
            {
                var line = await reader.ReadLineAsync();
                if (string.IsNullOrEmpty(line)) continue;
                
                if (line.StartsWith("data: "))
                {
                    var data = line.Substring(6);
                    try
                    {
                        var evt = JsonSerializer.Deserialize<StreamEvent>(data, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
                        if (evt != null)
                        {
                            yield return evt;
                            if (evt.Type == "done") yield break;
                        }
                    }
                    catch { }
                }
            }
        }

        public async Task<SkillInvokeResponse> InvokeSkillAsync(string skillId, object input)
        {
            return await PostAsync<SkillInvokeResponse>($"/api/skills/{skillId}/invoke", input);
        }

        // ============ Convenience Methods ============

        public async Task<ShellResponse> ExecuteShellAsync(string command, string cwd = null, int timeout = 60000)
        {
            var input = new { command, cwd, timeout };
            var result = await InvokeSkillAsync("shell_execute", input);
            return JsonSerializer.Deserialize<ShellResponse>(JsonSerializer.Serialize(result.Result), 
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        public async Task<PowerShellResponse> ExecutePowerShellAsync(string command, bool elevated = false, int timeout = 60000)
        {
            var input = new { command, elevated, timeout };
            var result = await InvokeSkillAsync("windows_powershell", input);
            return JsonSerializer.Deserialize<PowerShellResponse>(JsonSerializer.Serialize(result.Result),
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        public async Task<ProcessListResponse> ListProcessesAsync()
        {
            var input = new { action = "list" };
            var result = await InvokeSkillAsync("process_control", input);
            return JsonSerializer.Deserialize<ProcessListResponse>(JsonSerializer.Serialize(result.Result),
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        public async Task<SystemInfoResponse> GetSystemInfoAsync(string detail = "basic")
        {
            var input = new { detail };
            var result = await InvokeSkillAsync("system_info", input);
            return JsonSerializer.Deserialize<SystemInfoResponse>(JsonSerializer.Serialize(result.Result),
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        public async Task<WindowsServicesResponse> ListWindowsServicesAsync()
        {
            var input = new { action = "list" };
            var result = await InvokeSkillAsync("windows_services", input);
            return JsonSerializer.Deserialize<WindowsServicesResponse>(JsonSerializer.Serialize(result.Result),
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        public async Task<WindowsWingetResponse> SearchPackagesAsync(string package)
        {
            var input = new { action = "search", package };
            var result = await InvokeSkillAsync("windows_winget", input);
            return JsonSerializer.Deserialize<WindowsWingetResponse>(JsonSerializer.Serialize(result.Result),
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        public async Task<WindowsWingetResponse> InstallPackageAsync(string package, bool force = false)
        {
            var input = new { action = "install", package, acceptPackageAgreements = true, force };
            var result = await InvokeSkillAsync("windows_winget", input);
            return JsonSerializer.Deserialize<WindowsWingetResponse>(JsonSerializer.Serialize(result.Result),
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        public async Task<FileResponse> ReadFileAsync(string path)
        {
            var input = new { action = "read", target = path };
            var result = await InvokeSkillAsync("workspace_files", input);
            return JsonSerializer.Deserialize<FileResponse>(JsonSerializer.Serialize(result.Result),
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        public async Task<FileResponse> WriteFileAsync(string path, string content)
        {
            var input = new { action = "write", target = path, content };
            var result = await InvokeSkillAsync("workspace_files", input);
            return JsonSerializer.Deserialize<FileResponse>(JsonSerializer.Serialize(result.Result),
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        // ============ WebSocket API ============

        public async Task<bool> WsConnectAsync(Action onConnect = null, Action onDisconnect = null)
        {
            if (_wsConnected) return true;
            
            var wsUrl = _baseUrl.Replace("http://", "ws://").Replace("https://", "wss://") + "/ws";
            if (!string.IsNullOrEmpty(_token))
                wsUrl += $"?token={_token}";
            
            _ws = new ClientWebSocket();
            _wsCts = new CancellationTokenSource();
            
            try
            {
                await _ws.ConnectAsync(new Uri(wsUrl), _wsCts.Token);
                _wsConnected = true;
                
                _ = Task.Run(async () =>
                {
                    var buffer = new byte[8192];
                    while (_ws.State == WebSocketState.Open && !_wsCts.Token.IsCancellationRequested)
                    {
                        try
                        {
                            var result = await _ws.ReceiveAsync(new ArraySegment<byte>(buffer), _wsCts.Token);
                            if (result.MessageType == WebSocketMessageType.Text)
                            {
                                var json = Encoding.UTF8.GetString(buffer, 0, result.Count);
                                ProcessWsMessage(json);
                            }
                            else if (result.MessageType == WebSocketMessageType.Close)
                            {
                                _wsConnected = false;
                                onDisconnect?.Invoke();
                                break;
                            }
                        }
                        catch (OperationCanceledException) { break; }
                        catch { break; }
                    }
                });
                
                onConnect?.Invoke();
                return true;
            }
            catch
            {
                _wsConnected = false;
                return false;
            }
        }

        public async Task WsDisconnectAsync()
        {
            if (_ws != null && _ws.State == WebSocketState.Open)
            {
                _wsCts?.Cancel();
                await _ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closing", CancellationToken.None);
            }
            _ws?.Dispose();
            _ws = null;
            _wsConnected = false;
        }

        public void WsOn(string eventType, Action<object> callback)
        {
            if (_wsCallbacks.TryGetValue(eventType, out var list))
                list.Add(callback);
        }

        public void WsOff(string eventType, Action<object> callback)
        {
            if (_wsCallbacks.TryGetValue(eventType, out var list))
                list.Remove(callback);
        }

        public async Task WsSendAsync(string type, object payload = null)
        {
            if (!_wsConnected || _ws == null) throw new InvalidOperationException("WebSocket not connected");
            
            var msg = new Dictionary<string, object> { ["type"] = type };
            if (payload != null)
            {
                if (payload is IDictionary<string, object> dict)
                    foreach (var kvp in dict) msg[kvp.Key] = kvp.Value;
                else
                    msg["payload"] = payload;
            }
            
            var json = JsonSerializer.Serialize(msg);
            await _ws.SendAsync(new ArraySegment<byte>(Encoding.UTF8.GetBytes(json)), 
                WebSocketMessageType.Text, true, CancellationToken.None);
        }

        public async Task WsInitAsync()
        {
            await WsSendAsync("init");
        }

        public async Task WsPingAsync()
        {
            await WsSendAsync("ping");
        }

        public async Task WsRunAsync(string modelId, string objective, string sessionId = null,
            string[] enabledSkillIds = null, object settings = null)
        {
            await WsSendAsync("run", new { modelId, objective, sessionId, enabledSkillIds, settings });
        }

        public async Task<object> WsRunWithCallbackAsync(string modelId, string objective,
            Action<object> onProgress = null, Action<object> onDone = null,
            string sessionId = null, string[] enabledSkillIds = null, object settings = null,
            int timeoutMs = 120000)
        {
            var tcs = new TaskCompletionSource<object>();
            var cts = new CancellationTokenSource(timeoutMs);
            
            void ProgressHandler(object data)
            {
                onProgress?.Invoke(data);
            }
            
            void DoneHandler(object data)
            {
                onDone?.Invoke(data);
                tcs.TrySetResult(data);
            }
            
            WsOn("progress", ProgressHandler);
            WsOn("done", DoneHandler);
            
            cts.Token.Register(() => tcs.TrySetCanceled());
            
            await WsRunAsync(modelId, objective, sessionId, enabledSkillIds, settings);
            
            try
            {
                return await tcs.Task;
            }
            finally
            {
                WsOff("progress", ProgressHandler);
                WsOff("done", DoneHandler);
                cts.Dispose();
            }
        }

        private void ProcessWsMessage(string json)
        {
            try
            {
                using var doc = JsonDocument.Parse(json);
                var root = doc.RootElement;
                var type = root.GetProperty("type").GetString();
                
                if (type != null && _wsCallbacks.TryGetValue(type, out var callbacks))
                {
                    var payload = root.TryGetProperty("payload", out var p) ? p : root;
                    foreach (var cb in callbacks.ToArray())
                        cb.Invoke(payload);
                }
            }
            catch { }
        }

        // ============ HTTP Helpers ============

        private async Task<T> GetAsync<T>(string path)
        {
            var response = await _http.GetAsync($"{_baseUrl}{path}");
            response.EnsureSuccessStatusCode();
            var json = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<T>(json, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        private async Task<T> PostAsync<T>(string path, object body)
        {
            var json = JsonSerializer.Serialize(body);
            var content = new StringContent(json, Encoding.UTF8, "application/json");
            var response = await _http.PostAsync($"{_baseUrl}{path}", content);
            response.EnsureSuccessStatusCode();
            var responseJson = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<T>(responseJson, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }

        public void Dispose()
        {
            _wsCts?.Cancel();
            _ws?.Dispose();
            _http?.Dispose();
        }
    }

    // ============ Response Types ============

    public class HealthResponse { public bool Ok { get; set; } public string App { get; set; } }
    public class ModelsResponse { public Model[] Models { get; set; } }
    public class Model { public string Id { get; set; } public string Label { get; set; } public string Provider { get; set; } }
    public class SkillsResponse { public Skill[] Skills { get; set; } public string[] Warnings { get; set; } }
    public class Skill { public string Id { get; set; } public string Name { get; set; } public string Description { get; set; } }
    public class HooksResponse { public Hook[] Hooks { get; set; } }
    public class Hook { public string Id { get; set; } public string Description { get; set; } public string[] Events { get; set; } }
    public class SessionsResponse { public Session[] Sessions { get; set; } }
    public class SessionResponse { public Session Session { get; set; } }
    public class Session { public string Id { get; set; } public string Title { get; set; } public string UserId { get; set; } public object Settings { get; set; } }
    public class RunResponse { public string Answer { get; set; } public bool Done { get; set; } public string DoneReason { get; set; } public int Cycles { get; set; } public Session Session { get; set; } public object[] Trace { get; set; } }
    public class StreamEvent { public string Type { get; set; } public object Payload { get; set; } }
    public class SkillInvokeResponse { public bool Ok { get; set; } public string SkillId { get; set; } public object Result { get; set; } }
    public class ShellResponse { public bool Ok { get; set; } public int ExitCode { get; set; } public string Stdout { get; set; } public string Stderr { get; set; } public bool Blocked { get; set; } }
    public class PowerShellResponse { public bool Ok { get; set; } public int ExitCode { get; set; } public string Stdout { get; set; } public string Stderr { get; set; } public bool Elevated { get; set; } }
    public class ProcessListResponse { public bool Ok { get; set; } public string Action { get; set; } public int Count { get; set; } public ProcessInfo[] Processes { get; set; } }
    public class ProcessInfo { public string Name { get; set; } public int Pid { get; set; } }
    public class SystemInfoResponse { public bool Ok { get; set; } public string Hostname { get; set; } public string Platform { get; set; } public CpuInfo Cpu { get; set; } public MemoryInfo Memory { get; set; } }
    public class CpuInfo { public int Count { get; set; } public string Model { get; set; } }
    public class MemoryInfo { public long Total { get; set; } public long Free { get; set; } public long Used { get; set; } }
    public class WindowsServicesResponse { public bool Ok { get; set; } public int Count { get; set; } public WindowsService[] Services { get; set; } }
    public class WindowsService { public string Name { get; set; } public string DisplayName { get; set; } public string State { get; set; } }
    public class WindowsWingetResponse { public bool Ok { get; set; } public int Count { get; set; } public WingetPackage[] Packages { get; set; } }
    public class WingetPackage { public string Name { get; set; } public string Id { get; set; } public string Version { get; set; } }
    public class FileResponse { public bool Ok { get; set; } public string Path { get; set; } public string Content { get; set; } public bool Written { get; set; } }
}

File: clients/openclaw_client.py
#!/usr/bin/env python3
"""
OpenClaw Plus Python Client with WebSocket Support
Full REST + WebSocket client for game engines and applications.
"""

import json
import threading
import urllib.request
import urllib.error
from typing import Optional, Dict, Any, List, Callable
import time

try:
    import websocket

    WEBSOCKET_AVAILABLE = True
except ImportError:
    WEBSOCKET_AVAILABLE = False


class OpenClawClient:
    def __init__(
        self, base_url: str = "http://localhost:8787", token: Optional[str] = None
    ):
        self.base_url = base_url.rstrip("/")
        self.token = token
        self.headers = {"Content-Type": "application/json"}
        if token:
            self.headers["Authorization"] = f"Bearer {token}"

        self._ws = None
        self._ws_thread = None
        self._ws_connected = False
        self._ws_callbacks: Dict[str, List[Callable]] = {
            "hello": [],
            "init": [],
            "progress": [],
            "done": [],
            "error": [],
            "pong": [],
        }

    def _request(self, method: str, path: str, data: Optional[dict] = None) -> dict:
        url = f"{self.base_url}{path}"
        body = json.dumps(data).encode() if data else None
        req = urllib.request.Request(
            url, data=body, headers=self.headers, method=method
        )

        try:
            with urllib.request.urlopen(req, timeout=120) as response:
                return json.loads(response.read().decode())
        except urllib.error.HTTPError as e:
            error_body = e.read().decode() if e.fp else ""
            raise Exception(f"HTTP {e.code}: {error_body}")

    # ============ REST API ============

    def health(self) -> dict:
        return self._request("GET", "/api/health")

    def get_models(self) -> List[dict]:
        return self._request("GET", "/api/models").get("models", [])

    def get_skills(self) -> List[dict]:
        return self._request("GET", "/api/skills").get("skills", [])

    def get_hooks(self) -> List[dict]:
        return self._request("GET", "/api/hooks").get("hooks", [])

    def get_sessions(self) -> List[dict]:
        return self._request("GET", "/api/sessions").get("sessions", [])

    def get_session(self, session_id: str) -> dict:
        return self._request("GET", f"/api/sessions/{session_id}")

    def create_session(self, title: str, settings: Optional[dict] = None) -> dict:
        return self._request(
            "POST", "/api/sessions", {"title": title, "settings": settings}
        )

    def run(
        self,
        model_id: str,
        objective: str,
        session_id: Optional[str] = None,
        enabled_skill_ids: Optional[List[str]] = None,
        settings: Optional[dict] = None,
    ) -> dict:
        data = {
            "modelId": model_id,
            "objective": objective,
            "sessionId": session_id,
            "enabledSkillIds": enabled_skill_ids,
            "settings": settings,
        }
        return self._request("POST", "/api/run", data)

    def run_stream(
        self,
        model_id: str,
        objective: str,
        on_event: Callable[[dict], None],
        session_id: Optional[str] = None,
        enabled_skill_ids: Optional[List[str]] = None,
        settings: Optional[dict] = None,
        timeout_ms: int = 120000,
    ) -> dict:
        import http.client
        from urllib.parse import urlparse

        url = f"{self.base_url}/api/run/stream"
        parsed = urlparse(url)

        data = {
            "modelId": model_id,
            "objective": objective,
            "sessionId": session_id,
            "enabledSkillIds": enabled_skill_ids,
            "settings": settings,
        }

        body = json.dumps(data).encode()
        headers = {"Content-Type": "application/json"}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"

        if parsed.scheme == "https":
            conn = http.client.HTTPSConnection(
                parsed.netloc, timeout=timeout_ms // 1000
            )
        else:
            conn = http.client.HTTPConnection(parsed.netloc, timeout=timeout_ms // 1000)

        conn.request("POST", parsed.path, body, headers)
        response = conn.getresponse()

        final_result = None
        buffer = ""

        while True:
            chunk = response.read(1024).decode()
            if not chunk:
                break

            buffer += chunk
            lines = buffer.split("\n")
            buffer = lines.pop()

            for line in lines:
                if line.startswith("data: "):
                    try:
                        event_data = json.loads(line[6:])
                        if on_event:
                            on_event(event_data)
                        if event_data.get("type") == "done":
                            final_result = event_data
                    except json.JSONDecodeError:
                        pass

        conn.close()
        return final_result or {}

    def invoke_skill(self, skill_id: str, input_data: dict) -> dict:
        return self._request("POST", f"/api/skills/{skill_id}/invoke", input_data)

    # ============ Convenience Methods ============

    def shell_execute(
        self, command: str, cwd: Optional[str] = None, timeout: int = 60000
    ) -> dict:
        return self.invoke_skill(
            "shell_execute", {"command": command, "cwd": cwd, "timeout": timeout}
        )

    def powershell(
        self, command: str, elevated: bool = False, timeout: int = 60000
    ) -> dict:
        return self.invoke_skill(
            "windows_powershell",
            {"command": command, "elevated": elevated, "timeout": timeout},
        )

    def get_system_info(self, detail: str = "basic") -> dict:
        return self.invoke_skill("system_info", {"detail": detail})

    def list_processes(self) -> dict:
        return self.invoke_skill("process_control", {"action": "list"})

    def kill_process(self, target: str) -> dict:
        return self.invoke_skill(
            "process_control", {"action": "kill", "target": target}
        )

    def list_services(self) -> dict:
        return self.invoke_skill("windows_services", {"action": "list"})

    def start_service(self, name: str) -> dict:
        return self.invoke_skill("windows_services", {"action": "start", "name": name})

    def stop_service(self, name: str) -> dict:
        return self.invoke_skill("windows_services", {"action": "stop", "name": name})

    def search_packages(self, package: str) -> dict:
        return self.invoke_skill(
            "windows_winget", {"action": "search", "package": package}
        )

    def install_package(self, package: str, force: bool = False) -> dict:
        return self.invoke_skill(
            "windows_winget",
            {
                "action": "install",
                "package": package,
                "acceptPackageAgreements": True,
                "force": force,
            },
        )

    def read_file(self, path: str) -> dict:
        return self.invoke_skill("workspace_files", {"action": "read", "target": path})

    def write_file(self, path: str, content: str) -> dict:
        return self.invoke_skill(
            "workspace_files", {"action": "write", "target": path, "content": content}
        )

    def list_directory(self, path: str) -> dict:
        return self.invoke_skill("workspace_files", {"action": "list", "target": path})

    def get_network_adapters(self) -> dict:
        return self.invoke_skill("windows_network", {"action": "adapters"})

    def ping(self, host: str) -> dict:
        return self.invoke_skill("windows_network", {"action": "ping", "target": host})

    def get_event_logs(self, log: str = "Application", count: int = 20) -> dict:
        return self.invoke_skill(
            "windows_eventlog", {"action": "read", "log": log, "count": count}
        )

    def list_scheduled_tasks(self) -> dict:
        return self.invoke_skill("windows_tasks", {"action": "list"})

    def run_scheduled_task(self, path: str) -> dict:
        return self.invoke_skill("windows_tasks", {"action": "run", "path": path})

    # ============ WebSocket API ============

    def ws_connect(
        self,
        on_connect: Optional[Callable[[], None]] = None,
        on_disconnect: Optional[Callable[[], None]] = None,
    ) -> bool:
        if not WEBSOCKET_AVAILABLE:
            raise ImportError(
                "websocket-client not installed. Run: pip install websocket-client"
            )

        ws_url = (
            self.base_url.replace("http://", "ws://").replace("https://", "wss://")
            + "/ws"
        )
        if self.token:
            ws_url += f"?token={self.token}"

        def on_message(ws, message):
            try:
                data = json.loads(message)
                msg_type = data.get("type")
                if msg_type and msg_type in self._ws_callbacks:
                    for callback in self._ws_callbacks[msg_type]:
                        callback(data.get("payload", data))
            except json.JSONDecodeError:
                pass

        def on_error(ws, error):
            for callback in self._ws_callbacks.get("error", []):
                callback({"error": str(error)})

        def on_open(ws):
            self._ws_connected = True
            if on_connect:
                on_connect()

        def on_close(ws, close_status_code, close_msg):
            self._ws_connected = False
            if on_disconnect:
                on_disconnect()

        self._ws = websocket.WebSocketApp(
            ws_url,
            on_message=on_message,
            on_error=on_error,
            on_open=on_open,
            on_close=on_close,
        )

        self._ws_thread = threading.Thread(target=self._ws.run_forever, daemon=True)
        self._ws_thread.start()

        for _ in range(50):
            if self._ws_connected:
                return True
            time.sleep(0.1)

        return False

    def ws_disconnect(self):
        if self._ws:
            self._ws.close()
            self._ws = None
            self._ws_connected = False

    def ws_is_connected(self) -> bool:
        return self._ws_connected

    def ws_on(self, event_type: str, callback: Callable[[dict], None]):
        if event_type in self._ws_callbacks:
            self._ws_callbacks[event_type].append(callback)

    def ws_off(self, event_type: str, callback: Callable[[dict], None]):
        if event_type in self._ws_callbacks:
            if callback in self._ws_callbacks[event_type]:
                self._ws_callbacks[event_type].remove(callback)

    def ws_send(self, message_type: str, payload: dict = None):
        if not self._ws or not self._ws_connected:
            raise ConnectionError("WebSocket not connected")

        msg = {"type": message_type}
        if payload:
            msg.update(payload)

        self._ws.send(json.dumps(msg))

    def ws_init(self) -> None:
        self.ws_send("init")

    def ws_ping(self) -> None:
        self.ws_send("ping")

    def ws_run(
        self,
        model_id: str,
        objective: str,
        session_id: Optional[str] = None,
        enabled_skill_ids: Optional[List[str]] = None,
        settings: Optional[dict] = None,
    ):
        self.ws_send(
            "run",
            {
                "modelId": model_id,
                "objective": objective,
                "sessionId": session_id,
                "enabledSkillIds": enabled_skill_ids,
                "settings": settings,
            },
        )

    def ws_run_with_callback(
        self,
        model_id: str,
        objective: str,
        on_progress: Optional[Callable[[dict], None]] = None,
        on_done: Optional[Callable[[dict], None]] = None,
        session_id: Optional[str] = None,
        enabled_skill_ids: Optional[List[str]] = None,
        settings: Optional[dict] = None,
        timeout_ms: int = 120000,
    ) -> threading.Event:
        done_event = threading.Event()
        final_result = {"result": None}

        def on_progress_wrapper(data):
            if on_progress:
                on_progress(data)

        def on_done_wrapper(data):
            final_result["result"] = data
            done_event.set()
            if on_done:
                on_done(data)

        self.ws_on("progress", on_progress_wrapper)
        self.ws_on("done", on_done_wrapper)

        self.ws_run(model_id, objective, session_id, enabled_skill_ids, settings)

        def timeout_handler():
            time.sleep(timeout_ms / 1000)
            if not done_event.is_set():
                done_event.set()

        threading.Thread(target=timeout_handler, daemon=True).start()

        return done_event


if __name__ == "__main__":
    print("Testing OpenClaw REST API...")
    client = OpenClawClient()

    print("Health:", client.health())
    print("Models:", [m["id"] for m in client.get_models()])
    print("Skills:", [s["id"] for s in client.get_skills()])

    print("\nSystem Info:")
    info = client.get_system_info()
    if info.get("ok"):
        result = info.get("result", {})
        print(f"  Hostname: {result.get('hostname')}")
        print(f"  CPU: {result.get('cpu', {}).get('count')} cores")
        print(f"  Memory: {result.get('memory', {}).get('total', 0) // (1024**3)} GB")

    if WEBSOCKET_AVAILABLE:
        print("\nTesting WebSocket...")

        def on_hello(data):
            print(f"  Connected! Client ID: {data.get('clientId')}")

        def on_progress(data):
            print(f"  Progress: {data.get('type')}")

        def on_done(data):
            print(f"  Done: {data.get('done')}")

        client.ws_on("hello", on_hello)
        client.ws_on("progress", on_progress)
        client.ws_on("done", on_done)

        if client.ws_connect():
            print("  WebSocket connected")
            time.sleep(1)
            client.ws_disconnect()
            print("  WebSocket disconnected")
    else:
        print("\nWebSocket not available (install websocket-client)")
